src/wp-includes/class-wp-plugin-dependencies.php

```diff
@@ -0,0 +1,884 @@
+					continue;
+				}
+			}
+
+			if ( ! function_exists( 'plugins_api' ) ) {
+				require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+			}
+
+			$information = plugins_api(
+				'plugin_information',
+				array(
+					'slug'   => $slug,
+					'fields' => array(
+						'short_description' => true,
+						'icons'             => true,
+					),
+				)
+			);
+
+			// Ensure `self::$dependency_api_data` has data, sometimes resets after `plugins_api()`.
+			self::$dependency_api_data = (array) get_site_transient( 'wp_plugin_dependencies_plugin_data' );
+
+			if ( is_wp_error( $information ) ) {
+				continue;
+			}
+
+			// Ensure `self::$dependency_api_data` has data, sometimes resets after `plugins_api()`.
+			self::$dependency_api_data          = (array) get_site_transient( 'wp_plugin_dependencies_plugin_data' );
+			self::$dependency_api_data[ $slug ] = (array) $information;
+			// plugins_api() returns 'name' not 'Name'.
+			self::$dependency_api_data[ $information->slug ]['Name'] = self::$dependency_api_data[ $information->slug ]['name'];
+			set_site_transient( 'wp_plugin_dependencies_plugin_data', self::$dependency_api_data, 0 );
+		}
+
+		// Remove from self::$dependency_api_data if slug no longer a dependency.
+		$differences = array_diff( array_keys( self::$dependency_api_data ), self::$dependency_slugs );
+		foreach ( $differences as $difference ) {
+			unset( self::$dependency_api_data[ $difference ] );
+		}
+
+		ksort( self::$dependency_api_data );
+		// Remove empty elements.
+		self::$dependency_api_data = array_filter( self::$dependency_api_data );
+		set_site_transient( 'wp_plugin_dependencies_plugin_data', self::$dependency_api_data, 0 );
+	}
+
+	/**
+	 * Gets plugin directory names.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array An array of plugin directory names.
+	 */
+	protected static function get_plugin_dirnames() {
+		// Cache the plugin directory names.
+		if ( empty( self::$plugin_dirnames ) || self::$plugin_dirnames_cache !== self::$plugins ) {
+			self::$plugin_dirnames       = array();
+			self::$plugin_dirnames_cache = self::$plugins;
+
+			foreach ( array_keys( self::$plugins ) as $plugin ) {
+				$dirname = dirname( $plugin );
+
+				if ( '.' !== $dirname ) {
+					self::$plugin_dirnames[ $dirname ] = $plugin;
+				} else {
+					// Single file plugin.
+					self::$plugin_dirnames[ $plugin ] = $plugin;
+				}
+			}
+		}
+
+		return self::$plugin_dirnames;
+	}
+
+	/**
+	 * Gets circular dependency data.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array[] An array of circular dependency pairings.
+	 */
+	protected static function get_circular_dependencies() {
+		$circular_dependencies = array();
+		foreach ( self::$dependencies as $dependent => $dependencies ) {
+			/*
+			 * $dependent is in 'a/a.php' format. Dependencies are stored as slugs, i.e. 'a'.
+			 *
+			 * Convert $dependent to slug format for checking.
+			 */
+			$dependent_slug = dirname( $dependent );
```

Comment:
Let's add some protection for single file plugins.
```suggestion
			$dependent_slug = str_contains( $dependent, '/' ) ? dirname( $dependent ) : $dependent;
```